### DISCUSSION TITLE
Best practice for publication

### DISCUSSION THREAD

**Discussion by aTrotier** [MagneticResonanceImaging/MRIReco.jl#171] [General]
Hi,

I am finishing an article about T2 mapping with subspace and I want to share my sequence + the reconstruction. 

My idea is to create a dedicated package and with the binary for the Bruker sequence in the release like [this](https://github.com/aTrotier/SEQ_BRUKER_A_MP2RAGE_CS_PUBLIC/releases/tag/0.51) (I can't put the source code in public due to Bruker restriction)

The reconstruction is separated in 4 main parts :
- conversion from Bruker dataset to an acquisition data object (specific to the sequence)
- Generation of the subspace basis with the package [EPGsim.jl](https://github.com/MagneticResonanceImaging/EPGsim.jl) 
- Reconstruction with MRIReco ([already implemented earlier this year](https://magneticresonanceimaging.github.io/MRIReco.jl/latest/generated/examples/03-subspaceReconstruction/)) + BART for comparison.

I am writing the documentation with a literate example but I think the github action won't be able to perform the reconstruction.

Do you have some best practice to share the code related to an article ? (local module vs register package, managing scripts...)

@JakobAsslaender @tknopp @mrikasper @JeffFessler @felixhorger 

Thanks :)

**Comment by tknopp** [comment]
In my group, we have a dedicated organization for this purpose: https://github.com/IBIResearch
This is probably a good example: https://github.com/IBIResearch/IncreasedDynamicRange

The key thing there is that one needs to work with Projekt/Manifest.toml to really freeze the packages that have been used. We usually consider these repositories to be immutable and don't change them when the upstream packages change.

**Comment by JakobAsslaender** [comment]
For one less data-heavy paper I added all code to create the figures to the corresponding package's publication:
https://jakobasslaender.github.io/MRIgeneralizedBloch.jl/stable/

For our other publications, we either created a non-registered package (as we consider the code to be rather specialized):
https://github.com/JakobAsslaender/MRIeddyCurrentOptimization.jl

And for papers where there it is mostly scripts, we generally do a github w/o any package functionality. For data, I like https://zenodo.org

**Comment by felixhorger** [comment]
Thanks for reaching out! I agree with @tknopp and @JakobAsslaender 's suggestions.
On top, I can recommend to make the effort you put into structuring your code dependent on what you think its purpose will be.
Specifically, if you intend other researchers to use your sequence and pipeline beyond reproducing your results or if you are yourself interested in developing it further, it makes sense to put more effort into deciding what goes into a package and what can stay a script.

For example, assuming that the conversion bruker -> acq object doesn't change much over time, it might be a good idea to put it into a (unregistered) package. Depending on how your sequence works, the same might hold for the subspace basis generation.
I expect the reconstruction code to require maximum flexibility, hence I usually keep mine as scripts (the core functionality is already outsourced in MRIReco).

Hope that helps!

**Comment by aTrotier** [comment]
Thanks for the answer.

as @felixhorger suggests, I already put into an unregistered package the bruker->acq conversion, the basis and the fitting part (Ideally it should be linked to [this package ](https://github.com/aTrotier/QuantitativeMRI.jl) but I need to spend more time on it).

As you mentioned, the reconstruction part will be the one which might evolved a lot (I am using the v0.8 of MRIReco but in the next release, breaking changes has been implemented in [RegularizedLeastSquares.jl](https://github.com/JuliaImageRecon/RegularizedLeastSquares.jl/tree/master)). I will follow your advice and share a more simple script.

Do you think I should use an artefact.toml with zenodo (like @tknopp example) or just a path like in https://github.com/BRAIN-TO/GIRFReco.jl/blob/main/docs/lit/examples/joss_demo.jl ?

**Comment by aTrotier** [comment]
Thanks all for your answers !

The paper is currently in the review process with the corresponding repository : https://github.com/aTrotier/PAPER_subspace_MESE
It is a non-registered packaged.

For reproducibility : a literate example is available in the docs folder which reproduce the last figure of the paper. 
The user should download on zenodo the dataset, change the path of the script to the rawdata (and optionaly to BART). Then run the literate script with the `docs` environment in order to add others specific dependencies to create the figure (mainly CairoMakie)/

**Comment by aTrotier** [comment]
Thanks all, my paper is now available in MRM : https://onlinelibrary.wiley.com/doi/10.1002/mrm.30146
And the corresponding code has been moved to : https://github.com/CRMSB/PAPER_subspace_MESE
